chainer/links/loss/hierarchical_softmax.py

```diff
@@ -42,7 +54,23 @@ def _parse(self, node):
                 raise ValueError(
                     'All internal nodes must have two child nodes')
             left, right = node
+            if self.parent_index is not None:
+                lst = self.parent2child.get(self.parent_index, []) + [LEAF]
+                self.parent2child[self.parent_index] = lst
+
+            left_is_leaf = not isinstance(left, tuple)
+            right_is_leaf = not isinstance(right, tuple)
+            flag = left_is_leaf and not right_is_leaf
+            self.parent_index = self.next_id if flag else None
```

Comment:
This code is slightly hard to read. How about adding `parent_index` argument to `_parse` method?